File: nodejs/sample-app.js
#!/usr/bin/env node

/**
 * License Manager API - Node.js Sample Application.
 *
 * Usage: node sample-app.js
 * Requires: Node.js 18+ (uses built-in fetch)
 */

const readline = require('readline');
const { LicenseManagerClient } = require('./license-manager-client');

// ── Configuration ───────────────────────────────────────────────────────
const SERVER_URL = 'https://your-license-server.com';
const API_KEY = 'your-api-key-here';
const APP_URL = 'https://my-node-app.local';
const PRODUCT_ID = 'ABC12345';
const LICENSE_CODE = 'XXXX-XXXX-XXXX-XXXX';
const CLIENT_NAME = 'John Doe';

const client = new LicenseManagerClient({
  serverUrl: SERVER_URL,
  apiKey: API_KEY,
  applicationUrl: APP_URL,
});

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
const ask = (q) => new Promise((resolve) => rl.question(q, resolve));

function printResult(operation, result) {
  const status = result.is_active ? 'OK' : 'FAILED';
  console.log(`  [${status}] ${operation}: ${result.message || 'No message'}`);
}

async function main() {
  console.log('========================================');
  console.log('  License Manager - Node.js Example');
  console.log('========================================\n');

  while (true) {
    console.log('1. Check Connection');
    console.log('2. Activate License');
    console.log('3. Verify License');
    console.log('4. Deactivate License');
    console.log('5. Check for Updates');
    console.log('6. Download Update');
    console.log('7. Exit\n');

    const choice = (await ask('Choice: ')).trim();

    switch (choice) {
      case '1': {
        const result = await client.checkConnection();
        printResult('Connection', result);
        break;
      }
      case '2': {
        const result = await client.activateLicense(PRODUCT_ID, LICENSE_CODE, CLIENT_NAME);
        printResult('Activation', result);
        if (result.is_active) console.log('  License data saved to disk.');
        break;
      }
      case '3': {
        const result = await client.verifyLicense(PRODUCT_ID);
        printResult('Verification', result);
        break;
      }
      case '4': {
        const result = await client.deactivateLicense(PRODUCT_ID);
        printResult('Deactivation', result);
        break;
      }
      case '5': {
        const version = (await ask('Current version: ')).trim() || '1.0.0';
        const result = await client.checkForUpdate(PRODUCT_ID, version);
        printResult('Update Check', result);
        if (result.update_available) {
          console.log(`  New version: ${result.version}`);
          console.log(`  Update ID:   ${result.update_id}`);
          console.log(`  Summary:     ${result.summary}`);
        }
        break;
      }
      case '6': {
        const updateId = (await ask('Update ID (from update check): ')).trim();
        const dlType = (await ask('Type (main/sql) [main]: ')).trim() || 'main';
        try {
          const filePath = await client.downloadUpdate(updateId, '.', dlType);
          console.log(`  [OK] Downloaded: ${filePath}`);
        } catch (e) {
          console.log(`  [FAILED] Download: ${e.message}`);
        }
        break;
      }
      case '7':
        console.log('Goodbye!');
        rl.close();
        return;
      default:
        console.log('Invalid choice.');
    }

    console.log();
  }
}

main().catch(console.error);

File: python/sample_app.py
#!/usr/bin/env python3
"""
License Manager API - Python Sample Application.

Usage: pip install requests && python sample_app.py
"""

from license_manager_client import LicenseManagerClient

# ── Configuration ────────────────────────────────────────────────────────
SERVER_URL = "https://your-license-server.com"
API_KEY = "your-api-key-here"
APP_URL = "https://my-python-app.local"
PRODUCT_ID = "ABC12345"
LICENSE_CODE = "XXXX-XXXX-XXXX-XXXX"
CLIENT_NAME = "John Doe"


def main():
    client = LicenseManagerClient(
        server_url=SERVER_URL,
        api_key=API_KEY,
        application_url=APP_URL,
    )

    print("========================================")
    print("  License Manager - Python Example")
    print("========================================\n")

    while True:
        print("1. Check Connection")
        print("2. Activate License")
        print("3. Verify License")
        print("4. Deactivate License")
        print("5. Check for Updates")
        print("6. Download Update")
        print("7. Exit\n")

        choice = input("Choice: ").strip()

        if choice == "1":
            result = client.check_connection()
            print_result("Connection", result)

        elif choice == "2":
            result = client.activate_license(PRODUCT_ID, LICENSE_CODE, CLIENT_NAME)
            print_result("Activation", result)
            if result.get("is_active"):
                print("  License data saved to disk.")

        elif choice == "3":
            result = client.verify_license(PRODUCT_ID)
            print_result("Verification", result)

        elif choice == "4":
            result = client.deactivate_license(PRODUCT_ID)
            print_result("Deactivation", result)

        elif choice == "5":
            version = input("Current version: ").strip() or "1.0.0"
            result = client.check_for_update(PRODUCT_ID, version)
            print_result("Update Check", result)
            if result.get("update_available"):
                print(f"  New version: {result.get('version')}")
                print(f"  Update ID:   {result.get('update_id')}")
                print(f"  Summary:     {result.get('summary')}")

        elif choice == "6":
            update_id = input("Update ID (from update check): ").strip()
            dl_type = input("Type (main/sql) [main]: ").strip() or "main"
            try:
                file_path = client.download_update(update_id, ".", dl_type)
                print(f"  [OK] Downloaded: {file_path}")
            except Exception as e:
                print(f"  [FAILED] Download: {e}")

        elif choice == "7":
            print("Goodbye!")
            break

        else:
            print("Invalid choice.")

        print()


def print_result(operation: str, result: dict):
    status = "OK" if result.get("is_active") else "FAILED"
    print(f"  [{status}] {operation}: {result.get('message', 'No message')}")


if __name__ == "__main__":
    main()
